src/AppBundle/Service/JobHandler.php

```diff
@@ -213,20 +219,23 @@ private function getQuota(): int
     }
 
     /**
-     * Get the number of open connections to the replicas database.
+     * Get the number of open connections to the replicas databases.
      * @return int
      */
     private function getNumOpenConnections(): int
     {
+        $processCount = 0;
+        $registry = $this->container->get('doctrine');
         /** @var Connection $conn */
-        $conn = $this->container
-            ->get('doctrine')
-            ->getManager('replicas')
-            ->getConnection();
-
-        return (int)$conn->query(
-            'SELECT COUNT(*) FROM information_schema.PROCESSLIST'
-        )->fetchColumn(0);
+        foreach ($registry->getConnections() as $conn) {
+            if (!$conn->isConnected()) {
+                continue;
+            }
+            $processCount += (int)$conn->executeQuery(
+                'SELECT COUNT(*) FROM information_schema.PROCESSLIST'
+            )->fetchFirstColumn();
+        }
+        return $processCount;
```

Comment:
As I understand it, you can have 5 open connections to _any_ host, so this might incorrectly tell the system we're out of quota when we're not necessarily. I definitely, definitely do NOT think it's worth our time trying to figure out what connections we want to use when processing an event, and determining if there's quota only on those hosts, etc. The worse that will happen is it will force the event to be processed via the job queue, which is fine. Our low user base suggests we rarely if ever hit the connection limit anyway.

That said, I'm not sure about lines 231-233 That's skipping any connections we don't have an open connection to _in this process_, right? Do we know for sure that the connections we will need are already open at this point?

Reply:
You're right though, the way it's done is not correct! I didn't think it through properly.

I'm not sure what the best way is then, because surely if we connect to the 8 slices to check the connection count, we'll already be over the permitted 5? (By the way, where's that documented? I haven't been able to find details on wikitech.) Or no, it'd close each as we go wouldn't it?

Could we just set it to 5 and see how it goes?

Reply:
No I think we get 5 per host, so there's no risk in having connections open to all 8 simultaneously. Not sure if it's documented on Wikitech, and I can't even find it through codesearch either... I believe the limit is actually 10 (see old tasks like [T240036](https://phabricator.wikimedia.org/T240036) and [T177570](https://phabricator.wikimedia.org/T177570)), but we allow no more than 5 in Event Metrics to give some wiggle room for queries that aren't part of event processing.

This is pretty tricky, because we don't know which sections (aka slices) we need to connect to at this point in the code. It's a bit sloppy... but for now maybe just loop through s1 through s8 and return whichever has the highest value? That should effectively guard us from hitting the max connection limit, but at the cost of possibly making some jobs stay in a pending state longer than they need to (say s1 is full, but there's a job that only need s2).

Reply:
Okay, sounds good; I've done that.